File: backend/app/routers/currencies.py
"""Rotas de cotação de moedas: fiat (câmbio) + cripto, lista configurável pelo usuário."""

from __future__ import annotations

import secrets
from typing import Any

from fastapi import APIRouter, HTTPException, Query

from app.providers.currencies import (
    clean_crypto_code,
    clean_fiat_code,
    fetch_currency_quotes,
    mock_currencies_payload,
    search_crypto,
)
from app.schemas import (
    CurrenciesConfig,
    CurrenciesPatch,
    CurrenciesPayload,
    CurrencyItemBody,
    CurrencySearchResponse,
    OkResult,
)
from app.store import load, update

router = APIRouter(prefix="/api/currencies", tags=["currencies"])


def _config_public() -> CurrenciesConfig:
    cfg = load()
    return CurrenciesConfig.model_validate(cfg.get("currencies") or {})


@router.get("/config", response_model=CurrenciesConfig, summary="Configuração de cotação de moedas")
def get_config() -> CurrenciesConfig:
    return _config_public()


@router.patch("/config", response_model=CurrenciesConfig, summary="Atualiza cotação de moedas")
def patch_config(body: CurrenciesPatch) -> CurrenciesConfig:
    base_clean: str | None = None
    if body.base is not None:
        base_clean = clean_fiat_code(body.base)
        if not base_clean:
            raise HTTPException(400, "Moeda base inválida; use um código de 3 letras (ex.: BRL)")

    def mut(cfg: dict[str, Any]) -> None:
        cur = cfg.setdefault("currencies", {})
        if body.enabled is not None:
            cur["enabled"] = body.enabled
        if body.hidden is not None:
            cur["hidden"] = body.hidden
        if base_clean is not None:
            cur["base"] = base_clean

    update(mut)
    return _config_public()


@router.post("/items", response_model=CurrenciesConfig, summary="Adiciona uma moeda à lista")
def add_item(body: CurrencyItemBody) -> CurrenciesConfig:
    if body.kind == "fiat":
        code = clean_fiat_code(body.code)
        if not code:
            raise HTTPException(400, "Código de moeda inválido; use 3 letras (ex.: USD)")
    else:
        code = clean_crypto_code(body.code)
        if not code:
            raise HTTPException(400, "Identificador de criptomoeda inválido; escolha um resultado da busca")
    item_id = secrets.token_hex(4)

    def mut(cfg: dict[str, Any]) -> None:
        cur = cfg.setdefault("currencies", {})
        items = cur.setdefault("items", [])
        items.append({"id": item_id, "kind": body.kind, "code": code, "label": body.label.strip()})
        if not cur.get("enabled"):
            cur["enabled"] = True

    update(mut)
    return _config_public()


@router.delete("/items/{item_id}", response_model=OkResult, summary="Remove uma moeda da lista")
def delete_item(item_id: str) -> OkResult:
    def mut(cfg: dict[str, Any]) -> None:
        cur = cfg.setdefault("currencies", {})
        cur["items"] = [i for i in cur.get("items") or [] if i.get("id") != item_id]

    update(mut)
    return OkResult(ok=True)


@router.get("/search", response_model=CurrencySearchResponse, summary="Busca criptomoedas (proxy CoinGecko)")
def search(q: str = Query(..., min_length=2, description="Nome ou símbolo"), count: int = Query(8, ge=1, le=15)) -> CurrencySearchResponse:
    return CurrencySearchResponse(results=search_crypto(q, count))


@router.get("", response_model=CurrenciesPayload, summary="Cotações atuais (força fetch)")
def get_quotes() -> CurrenciesPayload:
    cfg = load()
    ccfg = cfg.get("currencies") or {}
    if cfg.get("mock") and ccfg.get("enabled"):
        return CurrenciesPayload.model_validate(mock_currencies_payload())
    data = fetch_currency_quotes(ccfg)
    return CurrenciesPayload.model_validate(data)

File: backend/src/routers/currencies.ts
import type { FastifyInstance } from "fastify";
import { randomBytes } from "node:crypto";
import {
  cleanCryptoCode,
  cleanFiatCode,
  fetchCurrencyQuotes,
  mockCurrenciesPayload,
  searchCrypto,
} from "../providers/currencies.js";
import { load, updateSync as update } from "../store.js";

export async function createCurrenciesRoutes(app: FastifyInstance): Promise<void> {
  app.get("/api/currencies/config", async () => {
    const cfg = load() as Record<string, unknown>;
    return (cfg.currencies ?? {}) as Record<string, unknown>;
  });

  app.patch("/api/currencies/config", async (request, reply) => {
    const body = request.body as Record<string, unknown> | null;
    if (!body) return reply.code(400).send({ ok: false, error: "corpo vazio" });
    let baseClean: string | null = null;
    if (body.base !== undefined && body.base !== null) {
      baseClean = cleanFiatCode(String(body.base));
      if (!baseClean) return reply.code(400).send({ ok: false, error: "Moeda base inválida; use um código de 3 letras (ex.: BRL)" });
    }
    update((cfg: Record<string, unknown>) => {
      const cur = (cfg.currencies ?? {}) as Record<string, unknown>;
      if (!cfg.currencies) cfg.currencies = cur;
      if (body.enabled !== undefined && body.enabled !== null) { cur.enabled = Boolean(body.enabled); cur.hidden = !Boolean(body.enabled); }
      else if (body.hidden !== undefined && body.hidden !== null) { cur.hidden = Boolean(body.hidden); cur.enabled = !Boolean(body.hidden); }
      if (baseClean !== null) cur.base = baseClean;
    });
    const cfg = load() as Record<string, unknown>;
    return (cfg.currencies ?? {}) as Record<string, unknown>;
  });

  app.post("/api/currencies/items", async (request, reply) => {
    const body = request.body as Record<string, unknown> | null;
    if (!body || typeof body.kind !== "string" || typeof body.code !== "string") return reply.code(400).send({ ok: false, error: "kind e code obrigatórios" });
    const kind = String(body.kind);
    let code: string | null;
    if (kind === "fiat") {
      code = cleanFiatCode(String(body.code));
      if (!code) return reply.code(400).send({ ok: false, error: "Código de moeda inválido; use 3 letras (ex.: USD)" });
    } else if (kind === "crypto") {
      code = cleanCryptoCode(String(body.code));
      if (!code) return reply.code(400).send({ ok: false, error: "Identificador de criptomoeda inválido; escolha um resultado da busca" });
    } else {
      return reply.code(400).send({ ok: false, error: "kind deve ser fiat ou crypto" });
    }
    const itemId = randomBytes(4).toString("hex");
    update((cfg: Record<string, unknown>) => {
      const cur = (cfg.currencies ?? {}) as Record<string, unknown>;
      if (!cfg.currencies) cfg.currencies = cur;
      const items = Array.isArray(cur.items) ? [...(cur.items as unknown[])] : [];
      items.push({ id: itemId, kind, code, label: String(body.label ?? "").trim() });
      cur.items = items;
      cur.enabled = true;
      cur.hidden = false;
    });
    const cfg = load() as Record<string, unknown>;
    return (cfg.currencies ?? {}) as Record<string, unknown>;
  });

  app.delete("/api/currencies/items/:item_id", async (request) => {
    const { item_id } = request.params as { item_id: string };
    update((cfg: Record<string, unknown>) => {
      const cur = (cfg.currencies ?? {}) as Record<string, unknown>;
      if (!cfg.currencies) cfg.currencies = cur;
      const items = Array.isArray(cur.items) ? (cur.items as Array<Record<string, unknown>>) : [];
      cur.items = items.filter((i) => String(i.id) !== item_id);
    });
    return { ok: true };
  });

  app.get("/api/currencies/search", async (request) => {
    const query = (request.query ?? {}) as Record<string, string>;
    const q = String(query.q ?? "");
    const count = Number(query.count ?? 8);
    const results = await searchCrypto(q, count);
    return { results };
  });

  app.get("/api/currencies", async () => {
    const cfg = load() as Record<string, unknown>;
    const ccfg = (cfg.currencies ?? {}) as Record<string, unknown>;
    if (cfg.mock && ccfg.enabled) return mockCurrenciesPayload();
    const data = await fetchCurrencyQuotes(ccfg);
    return data;
  });
}
